kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,248 @@
+//! Interface for reading, writing, and erasing storage blocks
+//! on devices without a Flash Translation Layer.
+//!
+//! Operates on discardable and writeable blocks, where areas must be
+//! discarded before overwriting.
+//!
+//! Here's an example implementation for a raw flash chip:
+//!
+//! ```rust,ignore
+//! use kernel::hil;
+//! use kernel::ErrorCode;
+//!
+//! const WRITE_BLOCK_BYTES: usize = 256;
+//! const DISCARD_BLOCK_BYTES: usize = 4096;
+//!
+//! struct RawFlashChip {};
+//!
+//! type WriteBlockIndex = hil::block_storage::BlockIndex<WRITE_BLOCK_BYTES>;
+//! type DiscardBlockIndex = hil::block_storage::BlockIndex<DISCARD_BLOCK_BYTES>;
+//!
+//! impl hil::block_storage::BlockStorage<WRITE_BLOCK_BYTES, DISCARD_BLOCK_BYTES> for RawFlashChip {
+//!     //(implement associated functions here)
+//! }
+//! ```
+use crate::ErrorCode;
+use core::ops::Add;
+
+/// An index to a block within device composed of blocks containing `B` bytes.
+/// Stores the number of blocks from the start of the device.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct BlockIndex<const B: usize>(pub u32);
+
+impl<const B: usize> BlockIndex<B> {
+    /// Returns the index that contains the address.
+    pub fn new_containing(byte_address: u64) -> Self {
```

Comment:
Why is this `u64`? Tock only supports 32 bit platforms.

Reply:
This is to support storage devices with more than 4GiB capacity.